src/clients/UBAClient/UBAClientUtilities.ts

```diff
@@ -77,3 +127,336 @@ export async function getUBAFeeConfig(
+        bundleEndBlockNumber: hubPoolClient.getLatestBundleEndBlockForChain(
+          relevantChainIds,
+          hubPoolBlockNumber,
+          chainId
+        ),
+        latestBlockNumber: spokePoolClient.latestBlockNumber,
+      },
+    };
+
+    for (const tokenSymbol of relevantTokenSymbols) {
+      // Get the bundles for this token
+      const availableBundles = chainState.bundles[tokenSymbol] ?? [];
+      // Get the block number and opening balance for this token
+      const { blockNumber, spokePoolBalance, incentiveBalance } = getOpeningTokenBalances(
+        chainId,
+        tokenSymbol,
+        hubPoolClient,
+        hubPoolBlockNumber
+      );
+      // Find the bundle that has the same block number as the current block number
+      const referenceBundleIndex = availableBundles.findLastIndex((bundle) => bundle.blockNumber === blockNumber);
+
+      const tokenMappingLookup = (
+        TOKEN_SYMBOLS_MAP as Record<string, { addresses: { [x: number]: string }; decimals: number }>
+      )[tokenSymbol];
+      const hubPoolTokenAddress = tokenMappingLookup.addresses[hubPoolClient.chainId];
+      const tokenDecimals = tokenMappingLookup.decimals;
+      const erc20 = ERC20__factory.connect(hubPoolTokenAddress, hubPoolClient.hubPool.provider);
+      const [hubBalance, hubEquity, ethSpokeBalance, spokeTargets] = await Promise.all([
+        erc20.balanceOf(hubPoolClient.hubPool.address, { blockTag: blockNumber }),
+        erc20.balanceOf(hubPoolClient.hubPool.address, { blockTag: blockNumber }),
+        erc20.balanceOf(spokePoolClient.spokePool.address, { blockTag: blockNumber }),
+        hubPoolClient.configStoreClient.getUBATargetSpokeBalances([chainId], hubPoolTokenAddress, blockNumber),
+      ]);
+
+      // Construct the bundle. If the bundle already exists, use the existing bundle
+      const constructedBundle: UBABundleState = {
+        ...(referenceBundleIndex !== -1 ? availableBundles[referenceBundleIndex] : { flows: [] }),
+        blockNumber,
+        openingBalance: spokePoolBalance,
+        openingIncentiveBalance: incentiveBalance,
+        config: {
+          ubaConfig: await getUBAFeeConfig(chainId, tokenSymbol, blockNumber),
+          tokenDecimals,
+          hubBalance,
+          hubEquity,
+          hubPoolSpokeBalance: ethSpokeBalance,
+          spokeTargets,
+        },
+      };
+      // If the bundle already exists, replace it
+      if (referenceBundleIndex !== -1) {
+        availableBundles[referenceBundleIndex] = constructedBundle;
+      }
+      // Otherwise, add it to the list of bundles
+      else {
+        availableBundles.push(constructedBundle);
+      }
+      // Find the last flow avaialble in the bundle
+      const lastFlow = constructedBundle.flows[constructedBundle.flows.length - 1];
+      // If the last flow exists, call getFlows with the last flow's block number as the fromBlock
+      const recentFlows = getFlows(
```

Comment:
Wouldn't you want to load all flows from the beginning of the bundle to the end, so wouldn't you want to set the `fromBlock` equal to the blockNumber of the first flow?

Reply:
Good point - answered below